File: Problem/Easy/1301-1400/1356. Sort Integers by The Number of 1 Bits/a.py
class Solution:
    def sortByBits(self, arr):
        # return sorted(arr, key=lambda x: [bin(x).count('1'), x]) # using array
        # using tuple
        return sorted(arr, key=lambda val: (bin(val).count('1'), val))


x = Solution().sortByBits(
    [1024, 512, 256, 128, 64, 32, 16, 8, 4, 2, 1, 8, 8, 7, 3, 9]
)

print(x)

File: Problem/Easy/1301-1400/1356. Sort Integers by The Number of 1 Bits/____a.java
// Comparator .... instead of lambda
// using Integer array
// 7 ms

import java.util.Arrays;
import java.util.Comparator;

class Solution {
	public int[] sortByBits(int[] arr) {
		// convert -> Integer
		Integer[] input = new Integer[arr.length];
		for (int i = 0; i < input.length; i++)
			input[i] = arr[i];

		// main logic
		Arrays.sort(input, new Comparator<Integer>() {
			@Override
			public int compare(Integer a, Integer b) {
				int a_bit = Integer.bitCount(a);
				int b_bit = Integer.bitCount(b);

				if (a_bit == b_bit)
					return a - b;

				return a_bit - b_bit;
			}
		});

		// convert -> int
		for (int i = 0; i < arr.length; i++)
			arr[i] = input[i];
		return arr;
	}
}
